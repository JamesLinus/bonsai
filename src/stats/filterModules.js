/**
 * @flow
 */

import type {ExtendedModule} from '../types/Stats';

export type FilterProps = {
  moduleName: string,
  cumulativeSizeMin: string,
  cumulativeSizeMax: string,
  requiredByCountMin: string,
  requiredByCountMax: string,
};

export type SortDirection = 'ASC' | 'DESC';
export type SortProps = {
  field: string,
  direction: SortDirection,
};

function makeRecordLikeRegExpFilter(field: string, re: RegExp | '') {
  return function(eModule: ExtendedModule) {
    if (re && !re.test(eModule[field])) {
      return false;
    }
    return true;
  };
}

function makeRecordRangeFilter(
  field: string,
  min: number | '',
  max: number | '',
) {
  return function(eModule: ExtendedModule) {
    if (min && max) {
      return eModule[field] >= min && eModule[field] <= max;
    }
    if (min) {
      return eModule[field] >= min;
    }
    if (max) {
      return eModule[field] <= max;
    }
    return true;
  };
}

export default function filterModules(
  extendedModules: Array<ExtendedModule>,
  filters: FilterProps,
  sort: SortProps,
) {
  return extendedModules
    .filter(makeRecordLikeRegExpFilter(
      'name',
      new RegExp(filters.moduleName),
    ))
    .filter(makeRecordRangeFilter(
      'cumulativeSize',
      Number(filters.cumulativeSizeMin),
      Number(filters.cumulativeSizeMax),
    ))
    .filter(makeRecordRangeFilter(
      'requiredByCount',
      Number(filters.requiredByCountMin),
      Number(filters.requiredByCountMax),
    ))
    .sort((a, b) => {
      const {field, direction} = sort;
      if (a[field] < b[field]) {
        return direction === 'ASC' ? -1 : 1;
      }
      if (a[field] > b[field]) {
        return direction === 'DESC' ? -1 : 1;
      }
      return 0;
    });
}
