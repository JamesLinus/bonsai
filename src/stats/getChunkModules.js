/*
 * @flow
 */

import type {RawStats, Module} from '../types/Stats';

import getModulesByChunk from './getModulesByChunk';
import getParentChunks from './getParentChunks';

export default function getChunkModules(
  stats: RawStats,
  selectedChunkId: number,
): ?Array<Module> {
  const parentChunks = getParentChunks(stats, selectedChunkId);
  if (!parentChunks) {
    return null;
  }

  const modulesByChunk = getModulesByChunk(stats);

  return parentChunks.reduce(
    (modules, chunk) => modules.concat(modulesByChunk[chunk.id].modules),
    [],
  );
}
