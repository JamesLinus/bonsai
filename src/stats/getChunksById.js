/*
 * @flow
 */

import type {Chunk} from '../types/Stats';

export type ChunksById = {[key: number]: Chunk};

export default function getChunksById(chunks: Array<Chunk>): ChunksById {
  const chunksById = {};
  chunks.forEach((module) => {
    chunksById[module.id] = module;
  });
  return chunksById;
}
