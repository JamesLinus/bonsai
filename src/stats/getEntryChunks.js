/*
 * @flow
 */

import type {RawStats} from '../types/Stats';

export default function getEntryChunks(stats: RawStats) {
  return stats.chunks;
  // .filter(
  //   chunk => chunk.names.length
  // );
}
