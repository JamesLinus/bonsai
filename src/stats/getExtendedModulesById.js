/*
 * @flow
 */

import type {RawStats, ExtendedModule} from '../types/Stats';

import getModulesByChunk from './getModulesByChunk';
import getParentChunkIds from './getParentChunkIds';

export type ExtendedModulesById = {[key: number]: ExtendedModule};

export default function getExtendedModulesById(
  stats: RawStats,
  selectedChunkId: number,
): ?ExtendedModulesById {
  const parentChunkIds = getParentChunkIds(
    stats,
    selectedChunkId
  );
  if (!parentChunkIds) {
    return null;
  }

  const modulesByChunk = getModulesByChunk(stats);

  const modules = parentChunkIds.reduce((modules, chunkId) => {
    return modules.concat(modulesByChunk[chunkId].modules);
  }, []);

  const extendedModulesById = modules.map((module) => {
    return {
      ...module,
      requirements: [],
    };
  }).reduce((map, module) => {
    map[module.id] = module;
    return map;
  }, {});

  modules.forEach((module) => {
    module.reasons.forEach((reason) => {
      // this module was added becuase $reason.
      // record the reason as an import on $reason.moduleId
      if (extendedModulesById[reason.moduleId]) {
        extendedModulesById[reason.moduleId].requirements.push(module);
      }
    });
  });

  return extendedModulesById;
}


