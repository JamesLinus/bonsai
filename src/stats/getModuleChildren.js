/*
 * @flow
 */

import type {ModulesById} from './getModulesById';



export default function getModuleChildren(modulesById: ModulesById, moduleId: number): Array<number> {
    return [];
}
