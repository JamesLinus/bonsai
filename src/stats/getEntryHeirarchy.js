/*
 * @flow
 */

import type {RawStats} from '../types/Stats';

import getEntryChunks from './getEntryChunks';

const localeChunkName = /(.*?)locale-[a-z]{2}(_[0-9A-Z]{2,3})?-(index|lite)(.*)/;

type Child = {
  id: number,
  ids: Array<number>,
  name: string,
  names: Array<string>,
  children: Array<Child>,
};

function squashLocaleChunks(name: string) {
  const match = name.match(localeChunkName);
  if (match) {
    return `${match[1]}local-.*-${match[3]}${match[4]}`;
  }
  return name;
}

function getChildrenForChunk(stats, parentChunk): Array<Child> {
  // collect chunks with the same name here
  const collectedChunksByName = {};

  getEntryChunks(stats).filter(
    (chunk) => chunk.parents.includes(parentChunk.id)
  ).map(
    (chunk) => ({
      id: chunk.id,
      ids: [chunk.id],
      name: chunk.names.map(squashLocaleChunks).join(', '),
      names: chunk.names,
      children: getChildrenForChunk(stats, chunk),
    })
  ).forEach((chunk) => {
    if (!collectedChunksByName[chunk.name]) {
      collectedChunksByName[chunk.name] = chunk;
    } else {
      collectedChunksByName[chunk.name].ids.push(chunk.ids);
    }
  });

  // $FlowFixMe: flow things `values` returns `Array<mixed>`
  const children: Array<Child> = Object.values(collectedChunksByName);
  return children;
}

export default function getEntryHeirarchy(stats: RawStats): {id: null, children: Array<Child>} {
  return {
    id: null,
    children: getEntryChunks(stats)
      .filter((chunk) => chunk.parents.length === 0)
      .map((chunk) => ({
        id: chunk.id,
        ids: [chunk.id],
        name: chunk.names.map(squashLocaleChunks).join(', '),
        names: chunk.names,
        children: getChildrenForChunk(stats, chunk),
      })),
  };
}
