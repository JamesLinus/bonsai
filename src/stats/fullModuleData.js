/*
 * @flow
 */

import type {
  ChunkID,
  ExtendedModule,
  ModuleID,
  RawStats,
} from '../types/Stats';

import type {Child} from './getEntryHeirarchy';

import getChunkModules from './getChunkModules';
import getEntryHeirarchy from './getEntryHeirarchy';
import getExtendedModulesById, {calculateModuleSizes} from './getExtendedModulesById';
import getModulesById from './getModulesById';
import getParentChunks from './getParentChunks';
import splitUnreachableModules from './splitUnreachableModules';

export default function fullModuleData(
  stats: RawStats,
  selectedChunkId: ?ChunkID,
  blacklistedModuleIds: Array<ModuleID>,
): {
  moduleData: ?{
    included: Array<ExtendedModule>,
    removed: Array<ExtendedModule>,
  },
  extendedModules: Array<ExtendedModule>,
  chunksByParent: Array<Child>,
  parentChunks: ?Array<Child>,
} {

  const chunksByParent = getEntryHeirarchy(stats).children;

  if (selectedChunkId === null || selectedChunkId === undefined) {
    return {
      moduleData: null,
      extendedModules: [],
      chunksByParent: chunksByParent,
      parentChunks: null,
    };
  }

  const parentChunks = getParentChunks(stats, selectedChunkId);
  const modules = getChunkModules(
    stats,
    parentChunks,
  );

  if (!modules) {
    return {
      moduleData: null,
      extendedModules: [],
      chunksByParent: chunksByParent,
      parentChunks: parentChunks,
    };
  }

  const extendedModulesById = getExtendedModulesById(modules);

  if (!extendedModulesById) {
    return {
      moduleData: null,
      extendedModules: [],
      chunksByParent: chunksByParent,
      parentChunks: parentChunks,
    };
  }

  const splitModules = splitUnreachableModules(
    extendedModulesById,
    blacklistedModuleIds,
  );

  // $FlowFixMe: flow thinks `values()` returns an `Array<mixed>` here
  const extendedModules: Array<ExtendedModule> = splitModules
    ? Object.values(calculateModuleSizes(getModulesById(splitModules.included)))
    : [];

  return {
    moduleData: splitModules,
    extendedModules: extendedModules,
    chunksByParent: chunksByParent,
    parentChunks: parentChunks,
  };
}
