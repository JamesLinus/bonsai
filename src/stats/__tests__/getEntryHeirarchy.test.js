/*
 * @flow
 */

import stats from '../../../public/data/stats-2017-03-13.json';

import getEntryHeirarchy from '../getEntryHeirarchy';

describe('getEntryHeirarchy', () => {
  it('should list all the chunks', () => {
    const result = getEntryHeirarchy(stats);

    expect(result).toMatchSnapshot();
  });
});
