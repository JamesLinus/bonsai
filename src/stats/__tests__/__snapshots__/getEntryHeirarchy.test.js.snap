exports[`getEntryHeirarchy should list all the chunks 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "id": 156,
                  "ids": Array [
                    156,
                  ],
                  "name": "CloseupContainer",
                  "names": Array [
                    "CloseupContainer",
                  ],
                },
                Object {
                  "children": Array [],
                  "id": 157,
                  "ids": Array [
                    157,
                  ],
                  "name": "webpack-dev",
                  "names": Array [
                    "webpack-dev",
                  ],
                },
                Object {
                  "children": Array [],
                  "id": 158,
                  "ids": Array [
                    158,
                  ],
                  "name": "UserSettingsPageContainer",
                  "names": Array [
                    "UserSettingsPageContainer",
                  ],
                },
                Object {
                  "children": Array [
                    Object {
                      "children": Array [],
                      "id": 156,
                      "ids": Array [
                        156,
                      ],
                      "name": "CloseupContainer",
                      "names": Array [
                        "CloseupContainer",
                      ],
                    },
                  ],
                  "id": 159,
                  "ids": Array [
                    159,
                  ],
                  "name": "AuthHomePage",
                  "names": Array [
                    "AuthHomePage",
                  ],
                },
                Object {
                  "children": Array [],
                  "id": 160,
                  "ids": Array [
                    160,
                  ],
                  "name": "SearchPageContainer",
                  "names": Array [
                    "SearchPageContainer",
                  ],
                },
                Object {
                  "children": Array [],
                  "id": 161,
                  "ids": Array [
                    161,
                  ],
                  "name": "BoardPage",
                  "names": Array [
                    "BoardPage",
                  ],
                },
                Object {
                  "children": Array [],
                  "id": 162,
                  "ids": Array [
                    162,
                  ],
                  "name": "BoardsListPage",
                  "names": Array [
                    "BoardsListPage",
                  ],
                },
                Object {
                  "children": Array [],
                  "id": 163,
                  "ids": Array [
                    163,
                  ],
                  "name": "ArticlePage",
                  "names": Array [
                    "ArticlePage",
                  ],
                },
                Object {
                  "children": Array [],
                  "id": 164,
                  "ids": Array [
                    164,
                  ],
                  "name": "ProfilePageContainer",
                  "names": Array [
                    "ProfilePageContainer",
                  ],
                },
                Object {
                  "children": Array [],
                  "id": 165,
                  "ids": Array [
                    165,
                  ],
                  "name": "ExplorePage",
                  "names": Array [
                    "ExplorePage",
                  ],
                },
                Object {
                  "children": Array [],
                  "id": 166,
                  "ids": Array [
                    166,
                  ],
                  "name": "UserProfilePinGridContainer",
                  "names": Array [
                    "UserProfilePinGridContainer",
                  ],
                },
                Object {
                  "children": Array [],
                  "id": 167,
                  "ids": Array [
                    167,
                  ],
                  "name": "NewsHubLandingPage",
                  "names": Array [
                    "NewsHubLandingPage",
                  ],
                },
                Object {
                  "children": Array [],
                  "id": 168,
                  "ids": Array [
                    168,
                  ],
                  "name": "DidItProfileFeedContainer",
                  "names": Array [
                    "DidItProfileFeedContainer",
                  ],
                },
                Object {
                  "children": Array [],
                  "id": 169,
                  "ids": Array [
                    169,
                  ],
                  "name": "UserLikesGridContainer",
                  "names": Array [
                    "UserLikesGridContainer",
                  ],
                },
              ],
              "id": 177,
              "ids": Array [
                177,
              ],
              "name": "entryChunk-www",
              "names": Array [
                "entryChunk-www",
              ],
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [],
                  "id": 155,
                  "ids": Array [
                    155,
                  ],
                  "name": "UnauthInterestFeedPageContainer",
                  "names": Array [
                    "UnauthInterestFeedPageContainer",
                  ],
                },
                Object {
                  "children": Array [],
                  "id": 171,
                  "ids": Array [
                    171,
                  ],
                  "name": "BenchmarkPage",
                  "names": Array [
                    "BenchmarkPage",
                  ],
                },
              ],
              "id": 191,
              "ids": Array [
                191,
              ],
              "name": "entryChunk-www-unauth",
              "names": Array [
                "entryChunk-www-unauth",
              ],
            },
            Object {
              "children": Array [],
              "id": 197,
              "ids": Array [
                197,
              ],
              "name": "external-core-1",
              "names": Array [
                "external-core-1",
              ],
            },
            Object {
              "children": Array [
                Object {
                  "children": Array [
                    Object {
                      "children": Array [
                        Object {
                          "children": Array [],
                          "id": 172,
                          "ids": Array [
                            172,
                          ],
                          "name": "entryChunk-react-analytics",
                          "names": Array [
                            "entryChunk-react-analytics",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 173,
                          "ids": Array [
                            173,
                          ],
                          "name": "entryChunk-react-sterling",
                          "names": Array [
                            "entryChunk-react-sterling",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 174,
                          "ids": Array [
                            174,
                          ],
                          "name": "entryChunk-react-eval",
                          "names": Array [
                            "entryChunk-react-eval",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 175,
                          "ids": Array [
                            175,
                          ],
                          "name": "entryChunk-react-common",
                          "names": Array [
                            "entryChunk-react-common",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 176,
                          "ids": Array [
                            176,
                          ],
                          "name": "entryChunk-react-PinBookmarklet",
                          "names": Array [
                            "entryChunk-react-PinBookmarklet",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 178,
                          "ids": Array [
                            178,
                          ],
                          "name": "entryChunk-react-PinBookmarkletReactWrapper",
                          "names": Array [
                            "entryChunk-react-PinBookmarkletReactWrapper",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 179,
                          "ids": Array [
                            179,
                          ],
                          "name": "entryChunk-react-OpenClosedInterestFeedReactPage",
                          "names": Array [
                            "entryChunk-react-OpenClosedInterestFeedReactPage",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 180,
                          "ids": Array [
                            180,
                          ],
                          "name": "entryChunk-react-UnauthOpenClosedPinReactPage",
                          "names": Array [
                            "entryChunk-react-UnauthOpenClosedPinReactPage",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 181,
                          "ids": Array [
                            181,
                          ],
                          "name": "entryChunk-react-UnauthOpenClosedBoardReactPage",
                          "names": Array [
                            "entryChunk-react-UnauthOpenClosedBoardReactPage",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 182,
                          "ids": Array [
                            182,
                          ],
                          "name": "entryChunk-react-OpenInterestFeedReactPage",
                          "names": Array [
                            "entryChunk-react-OpenInterestFeedReactPage",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 183,
                          "ids": Array [
                            183,
                          ],
                          "name": "entryChunk-react-UnauthOpenPinReactPage",
                          "names": Array [
                            "entryChunk-react-UnauthOpenPinReactPage",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 184,
                          "ids": Array [
                            184,
                          ],
                          "name": "entryChunk-react-UnauthOpenBoardReactPage",
                          "names": Array [
                            "entryChunk-react-UnauthOpenBoardReactPage",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 185,
                          "ids": Array [
                            185,
                          ],
                          "name": "entryChunk-react-InterestFeedReactPage",
                          "names": Array [
                            "entryChunk-react-InterestFeedReactPage",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 186,
                          "ids": Array [
                            186,
                          ],
                          "name": "entryChunk-react-UnauthBoardReactPage",
                          "names": Array [
                            "entryChunk-react-UnauthBoardReactPage",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 187,
                          "ids": Array [
                            187,
                          ],
                          "name": "entryChunk-react-UnauthPinReactPage",
                          "names": Array [
                            "entryChunk-react-UnauthPinReactPage",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 188,
                          "ids": Array [
                            188,
                          ],
                          "name": "entryChunk-react-UnauthHomeReactPage",
                          "names": Array [
                            "entryChunk-react-UnauthHomeReactPage",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 189,
                          "ids": Array [
                            189,
                          ],
                          "name": "entryChunk-react-UnauthLoginReactWrapper",
                          "names": Array [
                            "entryChunk-react-UnauthLoginReactWrapper",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 190,
                          "ids": Array [
                            190,
                          ],
                          "name": "entryChunk-react-UnauthSamsungCameraPinPage",
                          "names": Array [
                            "entryChunk-react-UnauthSamsungCameraPinPage",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 192,
                          "ids": Array [
                            192,
                          ],
                          "name": "external-analytics",
                          "names": Array [
                            "external-analytics",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 193,
                          "ids": Array [
                            193,
                          ],
                          "name": "external-sterling",
                          "names": Array [
                            "external-sterling",
                          ],
                        },
                        Object {
                          "children": Array [],
                          "id": 199,
                          "ids": Array [
                            199,
                          ],
                          "name": "Testing",
                          "names": Array [
                            "Testing",
                          ],
                        },
                      ],
                      "id": 194,
                      "ids": Array [
                        194,
                      ],
                      "name": "entryChunk-core",
                      "names": Array [
                        "entryChunk-core",
                      ],
                    },
                  ],
                  "id": 196,
                  "ids": Array [
                    196,
                  ],
                  "name": "external-core-2",
                  "names": Array [
                    "external-core-2",
                  ],
                },
              ],
              "id": 198,
              "ids": Array [
                198,
              ],
              "name": "external-core-1-slim",
              "names": Array [
                "external-core-1-slim",
              ],
            },
          ],
          "id": 195,
          "ids": Array [
            195,
          ],
          "name": "vendor-react",
          "names": Array [
            "vendor-react",
          ],
        },
      ],
      "id": 203,
      "ids": Array [
        203,
      ],
      "name": "scout",
      "names": Array [
        "scout",
      ],
    },
    Object {
      "children": Array [],
      "id": 204,
      "ids": Array [
        204,
      ],
      "name": "bundle-socialConnectPopup",
      "names": Array [
        "bundle-socialConnectPopup",
      ],
    },
  ],
  "id": -9007199254740991,
  "ids": Array [],
  "name": "",
  "names": Array [],
}
`;
